### ISSUE TITLE
Production import of Langmuir only processing ~1 work per hour; database errors in sidekiq

### ISSUE THREAD

**Issue by eporter23** [emory-libraries/dlp-curate#1042]
The most recent import of Langmuir has been running since 2/22. It has only processed 131 works in 4 days. Normally a batch of this size would have processed about 1000 works in this time.

Sidekiq has been showing over [50 retries](https://curate.library.emory.edu/sidekiq/retries) on the same error: 

ActiveRecord::StatementInvalid: Mysql2::Error::TimeoutError: Lock wait timeout exceeded; try restarting transaction: SELECT `minter_states`.* FROM `minter_states` WHERE `minter_states`.`namespace` = '...

Additional information about recent import attempts:
- 2/14: started last successful import (500 works). 
- 2/15: last reasonably successful import completed (500 works). The last work in the batch, however, is missing both of its filesets. [460qrfj6vv-cor](https://curate.library.emory.edu/concern/curate_generic_works/460qrfj6vv-cor?locale=en)
- 2/18: attempted to run final batch (1000 works). During this import job, we started seeing serious slowdown and the MySQL error noted above
- 2/19: stopped the batch started on 2/18 due to slowness
- 2/22: attempted to re-run the final batch (1000 works). Same behavior as we noted on 2/18. Only ~100 works have been created in 4 days.

**Comment by tmiles2** [comment]
Current steps to take: 
- [ ] Stop the current batch 
- [ ] Review what was deployed around the time of the issue

**Comment by devanshu-m** [comment]
Notes from @eporter23:

timeline of recent production imports
2/14: last reasonably successful import completed (500 works). The last work in the batch, however, is missing both of its filesets. 460qrfj6vv-cor
2/18: attempted to run final batch (1000 works). During this import job, we started seeing serious slowdown and MySQL errors
2/22: attempted to re-run the final batch (1000 works). Same behavior as we noted on 2/18. Only ~100 works have been created in 4 days.

**Comment by devanshu-m** [comment]
Notes from @eporter23:

timeline of recent production imports
2/14: last reasonably successful import completed (500 works). The last work in the batch, however, is missing both of its filesets. 460qrfj6vv-cor
2/18: attempted to run final batch (1000 works). During this import job, we started seeing serious slowdown and MySQL errors
2/22: attempted to re-run the final batch (1000 works). Same behavior as we noted on 2/18. Only ~100 works have been created in 4 days.

**Comment by devanshu-m** [comment]
1) Release deployed before last reasonably successful import (2/15) [last work missing file_sets]:

(2/13) v0.5.14:
Commits:
* Index thumbnail_path for Lux child works (#977)
* _ssi allows Lux to sort missing values last (#978)
* Sets the CORS header for the IIIF proxy (#982)
* When the Access-Control-Allow-Origin header is missing, the content
sent is not loaded by UV.
* Removes keyword arg from update_content (#968)
  * Removes preferred keyword from update_content method call in file_sets_controller. The preferred file symbol is being received as a normal arg in file_set_actor's update_content method and we were previously passing this a hashed key-value pair argument. By removing the key preferred: from the file_sets_controller, we pass this as a normal argument instead of the requirement of a keyword.


2) Release deployed after last reasonably successful import (2/15) [last work missing file_sets] and before first unsuccessful import (2/18) [serious slowdown and MySQL errors]:

(2/16) v0.5.15:
Commits in the deploy:
* Adds index fields of child works for Lux search results display of Parent works
* Adds a job to check if binaries exist in s3
* Adds preferred image into work display (#976)
* Adds access facet for lux (#998)

**Notes for this deploy:** I don't see anything in this deploy that would affect the minter state or database.

(2/18) v0.5.16:
Commits in the deploy:
* Adds request options to Faraday connection (#990)
  * Currently, request to create a graph for descendants in fedora times out. In this commit, we add request options when creating a Faraday connection. This way we can specify timeout in seconds so that the read request for graph finishes. This is added by default in ActiveFedora
12+ and this monkey patch can be removed once we upgrade AF. Refer: samvera/active_fedora#1271
  * Adds info about AF version
* Passes missing csv arg in check_binaries (#1003)
  * The csv was not being passed as an arg when running check_binaries job on all file_sets. This missing arg is being added to the method call and the definition. Tests are also updated for the same.

**Notes for this deploy:** I don't see anything in this deploy that would affect the minter state or database.

3) Release deployed after first unsuccessful import (2/18) [serious slowdown and MySQL errors] and before second unsuccessful import (2/22) [same slowdown behavior and MySQL errors]:

(2/19) v0.5.17:
Commits in the deploy:
* Index sort fields on collections (#1008)
* Manifests should always be available (#1009)
  * Move the IIIF manifest generation out of CurateGenericWorksController, where it is gated by Shibboleth, and into the iiif controller, where it is more publicly accessible. The goal is to let us load iiif images for public_low_view objects.
  * Refactor testing for manifest cacheing
  * Test the cache behavior against the ManifestBuilderService, which is where the behavior now actually lives.
  * Also, parsing the JSON and testing for specific tags is less likely to break than testing against a static fixture file.
* Redirects to origin when signing in via shib (#1004)
  * Redirect the user to omniauth request's origin if it exists, else will redirect to the dashboard. Refer: https://github.com/omniauth/omniauth/wiki/Saving-User-Location
* Adds a way to check if assign_id for works exists (#1011)
  * Modifies the assign_id method for the work, file_set, and collection.
  * Sometimes the minter goes into a locked state when it tries creating resources using IDs that are already taken or gone in Fedora. In order to make sure that a resource is only created on an unused valid ID, we loop through IDs until we find one that is not previously assigned to any resource in Fedora.

**Notes for this deploy:** I don't see anything in this deploy that would affect the database. There is, however, an update to the way IDs are being minted. This was successfully tested and resolved the issue where we used to see LDP::Conflict errors.

(2/20) v0.5.18:
Commits in the deploy:
* Change reading room label for Lux
* Changes manifest_url to meet lux requirements (#1014)
  * We are changing the manifest url to host_name/iiif/work.id/manifest to meet lux requirements.
In this commit, we are also changing the way build_manifest gets called. We now pass the presenter to the manifest builder class from the works_controller so we don't need to worry about building a presenter for the manifest builder service.
  * We are also editing the iiif_controller's manifest method when it calls the build_manifest method by passing a presenter that we create in the controller itself. Here, when we create a presenter for the work, we are passing the work's solr_doc, creating a new ability for the current_user, and the request is nil. When the request is nil, the manifest_url is being built by passing the host manually to the polymorphic_url method.

**Notes for this deploy:** I don't see anything in this deploy that would affect the minter state or database.

(2/21) v0.5.19:
Commits in the deploy:
* Index facetable fields for Lux (#1022)
* Changes ability for manifest creation to read all (#1023)
  * When we load a work in lux for which a manifest isn't cached, it creates a shortened manifest. If we then delete that manifest, and load the same work in curate, the manifest is generated properly.
  * Shortened manifest through lux might be because when lux calls the manifest method in the iiif_controller in curate (https://github.com/emory-libraries/dlp-curate/blob/master/app/controllers/iiif_controller.rb#L23), it creates an ability for the current_user (https://github.com/emory-libraries/dlp-curate/blob/master/app/controllers/iiif_controller.rb#L63). Since, lux might not know or pass the current_user, we get an incorrect manifest.

**Notes for this deploy:** I don't see anything in this deploy that would affect the minter state or database.

**Comment by rotated8** [comment]
I have talked to Jim from the DB team, and he is going to confirm with Solomon which database is causing the issue, and investigate.

**Comment by tmiles2** [comment]
Kathryn started a small batch of 10 objects so Jim could review the database during import.

**Comment by devanshu-m** [comment]
Notes:

* Yesterday, we tried minting an ID from the console manually and checking if it exists in Fedora. This was to make sure that the ID that gets minted is not locked and a resource can be created for newly minted IDs.

**Comment by eporter23** [comment]
On 2/28/20 we ran a 10 work Langmuir import in production with 2 members of the DBA team monitoring the database. From Mark Parten, 

> It looks like the innodb_lock_wait_timeout is set to 50 seconds.  This controls how long a lock request, such as a “select for update”, will wait to acquire the lock.  It may be that some of the transactions are acquiring the lock before the timeout occurs while others are not.  The developers will need to analyze the application to determine why these locks are occurring how best to resolve the locking issues you are now seeing.

**Comment by eporter23** [comment]
Also from Mark Parten: 

> It looks like session 15955 is the culprit.  I have seen several blocking issues and in all cases I’ve seen the blocking session was 15955.  Unfortunately we do not have a way to distinguish what this session is doing.  All of the effected sessions, as well as the problem session come through 10.65.89.92 almost certainly via the app.  The database is doing what it’s supposed to do.  When a row is locked by one session it doesn’t let another session lock it.

**Comment by eporter23** [comment]
In sprint planning 3/2/20 we identified the following as next steps:

The minter pre-check which was added several weeks ago seems to be exceeding 50 seconds to check Fedora. If we try to run this check against SOLR instead of Fedora, we won't be able to include deleted objects (which are removed fro the index) which may still exist in Fedora. Later on, we may want to revisit this issue to see if anything can be done about the 50 second timeout.

- [ ] Remove the minter pre-check which was added several weeks ago, but it seems to be exceeding 50 seconds to check Fedora
- [ ] Run a small test load in curate-test
- [ ] Assess performance time
- [ ] If performance is improved, remove the minter pre-check from all environments until production data loads are complete

If we run into further minter conflicts, we can run a utility to resolve this.
